### PR TITLE
Wise: getToken mutex

### PR DESCRIPTION
**Pull request by kewitz**
I'm currently investigating a series of failed Wise authorizations related to refresh token requests.
As a first step, I'm following the recommendation on Wise documentation and making sure there's no racing condition between these calls.

Reference: https://open-collective.sentry.io/issues/3490446877/events/a97a4dd3ef204142aeabef2d45877f6d/events/?project=5199682

### CODE REVIEW

File: server/lib/mutex.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Cache based, service-wide mutex implementation.
+ */
+export async function lockUntilResolved<T>(
+  key: string,
+  until: () => Promise<T>,
+  { lockAcquireTimeoutMs = 15 * 1000, unlockTimeoutMs = 60 * 1000 } = {},
+): Promise<T> {
+  const redis = await createRedisClient();
```
Comment (znarf) [inline server/lib/mutex.ts:17]: Have you considered targeting another specific redis instance?

File: server/lib/mutex.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  if (!redis) {
+    logger.warn(`Redis is not available, ${key} running without a mutex lock!`);
+    return until();
+  }
+  const _key = `lock:${key}`;
+  const start = Date.now();
+  let lockAcquired = await redis.mSetNX([_key, '1']);
```
[the remark this answers is not in the document]
Comment (kewitz) [reply server/lib/mutex.ts:24]: I wanted it to be atomic.

[the remark this answers is not in the document]
Comment (znarf) [reply server/lib/mutex.ts:24]: There is nothing particularly atomic when using it to set a single value? It's atomic when setting multiple value.

A set should be enough here.

File: server/lib/mutex.ts (hunk 1 above)
Comment (znarf) [reply server/lib/mutex.ts:17]: We've seen keys being flushed very quickly.

It **may** work.

File: server/lib/mutex.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Betree) [reply server/lib/mutex.ts:24]: I did not know `mSetNX`, but looking at the docs it does look like it is atomic - even with a single value:

image

[the remark this answers is not in the document]
Comment (znarf) [reply server/lib/mutex.ts:24]: Ok, so what we're looking for here is "will not perform any operation at all even if just a single key already exists".

We want to do `get + set (if null)` in an atomic way.

[the remark this answers is not in the document]
Comment (znarf) [reply server/lib/mutex.ts:24]: The proper method is https://redis.io/commands/setnx/ (deprecated). The new recommended way is https://redis.io/commands/set/ with a `NX` flag.

File: server/lib/mutex.ts (hunk 1 above)
Comment (kewitz) [reply server/lib/mutex.ts:17]: You're correct. I think the nature of the other cache is very different from this one. Those are discardable and these are not.

**Comment by kewitz** [comment]
Thanks for the suggestions and details @znarf and @Betree. Let me know if you have any pending observations.

**Review by Betree — APPROVED** [review]
The lock function looks good to me

**Comment by Betree** [comment]
@kewitz The test for the mutex randomly fails ([example](https://github.com/opencollective/opencollective-api/actions/runs/8108579900/job/22162051012)). I recommend implementing one of the solutions proposed in https://github.com/opencollective/opencollective-api/pull/9856#discussion_r1500300392, or setting more stable values.

image

**Pull request merged**